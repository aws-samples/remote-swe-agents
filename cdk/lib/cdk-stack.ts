import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SlackBolt } from './constructs/slack-bolt';
import { Worker } from './constructs/worker';
import { SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import { Storage } from './constructs/storage';
import { EC2GarbageCollector } from './constructs/ec2-gc';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { BlockPublicAccess, Bucket, ObjectOwnership } from 'aws-cdk-lib/aws-s3';
import { RemovalPolicy } from 'aws-cdk-lib';
import { EdgeFunction } from './constructs/cf-lambda-furl-service/edge-function';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { HostedZone } from 'aws-cdk-lib/aws-route53';
import { Auth } from './constructs/auth';
import { AsyncJob } from './constructs/async-job';
import { Webapp } from './constructs/webapp';

export interface MainStackProps extends cdk.StackProps {
  readonly signPayloadHandler: EdgeFunction;
  readonly domainName?: string;
  readonly sharedCertificate?: ICertificate;
  readonly cloudFrontWebAclArn?: string;
  readonly vpcId?: string;

  readonly slack: {
    botTokenParameterName: string;
    signingSecretParameterName: string;
    adminUserIdList?: string;
  };
  readonly github:
    | {
        appId: string;
        installationId: string;
        privateKeyParameterName: string;
      }
    | {
        personalAccessTokenParameterName: string;
      };
  readonly loadBalancing?: {
    awsAccounts: string[];
    roleName: string;
  };
  readonly workerAmiIdParameterName: string;
  readonly additionalAwsManagedPolicies?: string[];
  readonly initialWebappUserEmail?: string;
}

export class MainStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: MainStackProps) {
    super(scope, id, { ...props, description: `${props.description ?? 'Remote SWE Agents stack'} (uksb-lv52f92xel)` });

    const botToken = StringParameter.fromStringParameterAttributes(this, 'SlackBotToken', {
      parameterName: props.slack.botTokenParameterName,
      forceDynamicReference: true,
    });

    const signingSecret = StringParameter.fromStringParameterAttributes(this, 'SlackSigningSecret', {
      parameterName: props.slack.signingSecretParameterName,
      forceDynamicReference: true,
    });

    const workerAmiIdParameter = StringParameter.fromStringParameterAttributes(this, 'WorkerAmiId', {
      parameterName: props.workerAmiIdParameterName,
      forceDynamicReference: true,
    });

    const hostedZone = props.domainName
      ? HostedZone.fromLookup(this, 'HostedZone', {
          domainName: props.domainName,
        })
      : undefined;

    const accessLogBucket = new Bucket(this, 'AccessLog', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      enforceSSL: true,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
    });

    // 既存VPCが指定されていればそれを使用し、そうでなければ新規作成
    const vpc = props.vpcId
      ? Vpc.fromLookup(this, 'VpcV2', { vpcId: props.vpcId })
      : new Vpc(this, 'VpcV2', {
          subnetConfiguration: [
            {
              // We use public subnets for worker EC2 instances. Here are the reasons:
              //   1. to remove NAT Gateway cost
              //   2. to avoid IP-address-based throttling/filtering from external services
              // All the instances are securely protected by security groups without any inbound rules.
              subnetType: SubnetType.PUBLIC,
              name: 'Public',
              cidrMask: 20,
            },
          ],
        });

    const storage = new Storage(this, 'Storage', { accessLogBucket });

    const worker = new Worker(this, 'Worker', {
      vpc,
      storageTable: storage.table,
      imageBucket: storage.bucket,
      slackBotTokenParameter: botToken,
      ...('appId' in props.github
        ? {
            gitHubApp: {
              appId: props.github.appId,
              installationId: props.github.installationId,
              privateKeyParameterName: props.github.privateKeyParameterName,
            },
          }
        : {
            githubPersonalAccessTokenParameter: StringParameter.fromStringParameterAttributes(
              this,
              'GitHubPersonalAccessToken',
              {
                parameterName: props.github.personalAccessTokenParameterName,
                forceDynamicReference: true,
              }
            ),
          }),
      loadBalancing: props.loadBalancing,
      accessLogBucket,
      amiIdParameterName: workerAmiIdParameter.parameterName,
      additionalAwsManagedPolicies: props.additionalAwsManagedPolicies,
    });

    new SlackBolt(this, 'SlackBolt', {
      botTokenParameter: botToken,
      signingSecretParameter: signingSecret,
      launchTemplateId: worker.launchTemplate.launchTemplateId!,
      subnetIdListForWorkers: vpc.publicSubnets.map((s) => s.subnetId).join(','),
      workerBus: worker.bus,
      storage,
      adminUserIdList: props.slack.adminUserIdList,
      workerLogGroupName: worker.logGroup.logGroupName,
      workerAmiIdParameter,
    });

    new EC2GarbageCollector(this, 'EC2GarbageCollector', {
      expirationInDays: 1,
      imageRecipeName: worker.imageBuilder.imageRecipeName,
    });

    const auth = new Auth(this, 'Auth', {
      hostedZone,
      sharedCertificate: props.sharedCertificate,
      initialUserEmail: props.initialWebappUserEmail,
    });

    worker.bus.addUserPoolProvider(auth.userPool);

    const asyncJob = new AsyncJob(this, 'AsyncJob', { storage });

    const webapp = new Webapp(this, 'Webapp', {
      storage,
      hostedZone,
      certificate: props.sharedCertificate,
      signPayloadHandler: props.signPayloadHandler,
      webAclArn: props.cloudFrontWebAclArn,
      accessLogBucket,
      auth,
      launchTemplateId: worker.launchTemplate.launchTemplateId!,
      subnetIdListForWorkers: vpc.publicSubnets.map((s) => s.subnetId).join(','),
      workerBus: worker.bus,
      asyncJob,
      workerAmiIdParameter,
    });

    new cdk.CfnOutput(this, 'FrontendDomainName', {
      value: webapp.baseUrl,
    });
  }
}
