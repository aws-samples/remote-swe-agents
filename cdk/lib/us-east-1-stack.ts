import * as cdk from 'aws-cdk-lib';
import { Certificate, CertificateValidation, ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ARecord, HostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';
import { EdgeFunction } from './constructs/cf-lambda-furl-service/edge-function';
import { CommonWebAcl } from './constructs/web-acl';
import { join } from 'path';
import { IRepository, Repository } from 'aws-cdk-lib/aws-ecr';

interface UsEast1StackProps extends cdk.StackProps {
  domainName?: string;
  allowedIpV4AddressRanges?: string[];
  allowedIpV6AddressRanges?: string[];
  allowedCountryCodes?: string[];
  enableAgentCore: boolean;
}

export class UsEast1Stack extends cdk.Stack {
  /**
   * the ACM certificate for CloudFront (it must be deployed in us-east-1).
   * undefined if domainName is not set.
   */
  public readonly certificate: ICertificate | undefined = undefined;
  /**
   * the signer L@E function (it must be deployed in us-east-1).
   */
  public readonly signPayloadHandler: EdgeFunction;
  /**
   * the WAF Web ACL ARN for CloudFront (it must be deployed in us-east-1).
   * undefined if no IP restrictions are set.
   */
  public readonly webAclArn: string | undefined = undefined;

  public readonly agentCoreRepository?: IRepository;

  constructor(scope: Construct, id: string, props: UsEast1StackProps) {
    super(scope, id, props);

    if (props.domainName) {
      const hostedZone = HostedZone.fromLookup(this, 'HostedZone', {
        domainName: props.domainName,
      });

      // cognito requires A record for Hosted UI custom domain
      // https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html#cognito-user-pools-add-custom-domain-adding
      // > Its parent domain must have a valid DNS A record. You can assign any value to this record.
      new ARecord(this, 'Record', {
        zone: hostedZone,
        target: RecordTarget.fromIpAddresses('8.8.8.8'),
      });

      const cert = new Certificate(this, 'CertificateV2', {
        domainName: `*.${hostedZone.zoneName}`,
        validation: CertificateValidation.fromDns(hostedZone),
        subjectAlternativeNames: [hostedZone.zoneName],
      });
      this.certificate = cert;
    }

    const signPayloadHandler = new EdgeFunction(this, 'SignPayloadHandler', {
      entryPath: join(__dirname, 'constructs', 'cf-lambda-furl-service', 'lambda', 'sign-payload.ts'),
    });

    this.signPayloadHandler = signPayloadHandler;

    if (props.allowedIpV4AddressRanges || props.allowedIpV6AddressRanges || props.allowedCountryCodes) {
      const webAcl = new CommonWebAcl(this, 'WebAcl', {
        scope: 'CLOUDFRONT',
        allowedIpV4AddressRanges: props.allowedIpV4AddressRanges,
        allowedIpV6AddressRanges: props.allowedIpV6AddressRanges,
        allowedCountryCodes: props.allowedCountryCodes,
      });

      this.webAclArn = webAcl.webAclArn;
    }

    if (props.enableAgentCore) {
      const parent = new Construct(this, 'AgentCoreRepository');
      const repositoryName = cdk.Names.uniqueResourceName(parent, { maxLength: 64 }).toLowerCase();
      new Repository(parent, 'Resource', {
        repositoryName,
      });
      this.agentCoreRepository = Repository.fromRepositoryName(this, 'AgentCoreRepositoryReference', repositoryName);
    }
  }
}
