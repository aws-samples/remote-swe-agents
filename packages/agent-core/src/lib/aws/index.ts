export * from './s3';
export * from './ddb';
export * from './ec2';
export * from './ssm';
export * from '../config';
