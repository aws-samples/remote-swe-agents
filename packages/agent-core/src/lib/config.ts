import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand, PutCommand } from '@aws-sdk/lib-dynamodb';
import { CONFIG_TABLE_NAME, MODEL_CONFIG_KEY } from '../schema/config';

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const getModelConfig = async (): Promise<string | undefined> => {
  const tableName = process.env.CONFIG_TABLE_NAME || CONFIG_TABLE_NAME;

  try {
    const command = new GetCommand({
      TableName: tableName,
      Key: {
        PK: MODEL_CONFIG_KEY,
        SK: MODEL_CONFIG_KEY,
      },
    });

    const response = await docClient.send(command);
    return response.Item?.value;
  } catch (error) {
    console.error('Error getting model config:', error);
    return undefined;
  }
};

export const setModelConfig = async (modelId: string): Promise<void> => {
  const tableName = process.env.CONFIG_TABLE_NAME || CONFIG_TABLE_NAME;

  try {
    const command = new PutCommand({
      TableName: tableName,
      Item: {
        PK: MODEL_CONFIG_KEY,
        SK: MODEL_CONFIG_KEY,
        value: modelId,
      },
    });

    await docClient.send(command);
  } catch (error) {
    console.error('Error setting model config:', error);
    throw error;
  }
};
