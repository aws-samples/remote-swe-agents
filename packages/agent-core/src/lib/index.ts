export * from './kill-timer';
export * from './prompt';
export * from './slack';
export * from './messages';
export * from './metadata';
export * from './prompt';
