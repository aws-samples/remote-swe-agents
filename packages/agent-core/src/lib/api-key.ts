import { GetCommand, PutCommand, QueryCommand, DeleteCommand } from '@aws-sdk/lib-dynamodb';
import { ddb, TableName } from './aws';
import { ApiKeyItem } from '../schema';
import crypto from 'crypto';

/**
 * Create a new API key
 * @param description Optional description for the key
 * @param ownerId Optional owner ID
 * @returns The generated API key
 */
export const createApiKey = async (description?: string, ownerId?: string): Promise<string> => {
  const now = Date.now();
  const timestamp = String(now).padStart(15, '0');

  // Generate a random 32 byte key and hex encode it
  const apiKey = crypto.randomBytes(32).toString('hex');

  await ddb.send(
    new PutCommand({
      TableName,
      Item: {
        PK: 'api-key',
        SK: apiKey,
        LSI1: timestamp,
        createdAt: now,
        description,
        ownerId,
      } satisfies ApiKeyItem,
    })
  );

  return apiKey;
};

/**
 * Validate if an API key exists
 * @param apiKey The API key to validate
 * @returns true if the key exists, false otherwise
 */
export const validateApiKey = async (apiKey: string): Promise<boolean> => {
  const result = await ddb.send(
    new GetCommand({
      TableName,
      Key: {
        PK: 'api-key',
        SK: apiKey,
      },
    })
  );

  return !!result.Item;
};

/**
 * Get all API keys
 * @param limit Maximum number of keys to return
 * @returns Array of API key items
 */
export const getApiKeys = async (limit: number = 50): Promise<ApiKeyItem[]> => {
  const res = await ddb.send(
    new QueryCommand({
      TableName,
      IndexName: 'LSI1',
      KeyConditionExpression: 'PK = :pk',
      ExpressionAttributeValues: {
        ':pk': 'api-key',
      },
      ScanIndexForward: false, // DESC order
      Limit: limit,
    })
  );

  return (res.Items ?? []) as ApiKeyItem[];
};

/**
 * Delete an API key
 * @param apiKey The API key to delete
 */
export const deleteApiKey = async (apiKey: string): Promise<void> => {
  await ddb.send(
    new DeleteCommand({
      TableName,
      Key: {
        PK: 'api-key',
        SK: apiKey,
      },
    })
  );
};
