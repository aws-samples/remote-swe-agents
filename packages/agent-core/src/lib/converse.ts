import {
  BedrockRuntimeClient,
  ConverseCommand,
  ConverseCommandInput,
  ConverseResponse,
  ThrottlingException,
} from '@aws-sdk/client-bedrock-runtime';
import { AssumeRoleCommand, STSClient } from '@aws-sdk/client-sts';
import { ddb, TableName } from './aws';
import { GetCommand, PutCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { z } from 'zod';

const sts = new STSClient();
const awsAccounts = (process.env.BEDROCK_AWS_ACCOUNTS ?? '').split(',');
const roleName = process.env.BEDROCK_AWS_ROLE_NAME || 'bedrock-remote-swe-role';

// State management for persistent account selection and retry
let currentAccountIndex = 0; // Currently used account index

// Keywords for thinking budget adjustment
const ULTRA_THINKING_KEYWORD = 'ultrathink';

const defaultOutputTokenCount = 8192;

const modelTypeSchema = z.enum(['sonnet3.5v1', 'sonnet3.5', 'sonnet3.7', 'haiku3.5', 'nova-pro', 'opus4', 'sonnet4']);
type ModelType = z.infer<typeof modelTypeSchema>;

const modelConfigSchema = z.object({
  maxOutputTokens: z.number().default(4096),
  maxInputTokens: z.number(),
  cacheSupport: z.array(z.enum(['system', 'tool', 'message'])).default([]),
  reasoningSupport: z.boolean().default(false),
  toolChoiceSupport: z.array(z.enum(['any', 'auto', 'tool'])).default([]),
});

const modelConfigs: Record<ModelType, Partial<z.infer<typeof modelConfigSchema>>> = {
  'sonnet3.5v1': {
    maxOutputTokens: 4096,
    maxInputTokens: 200_000,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
  'sonnet3.5': {
    maxOutputTokens: 4096,
    maxInputTokens: 200_000,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
  'sonnet3.7': {
    maxOutputTokens: 64_000,
    maxInputTokens: 200_000,
    cacheSupport: ['system', 'message', 'tool'],
    reasoningSupport: true,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
  'haiku3.5': {
    maxOutputTokens: 4096,
    maxInputTokens: 200_000,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
  'nova-pro': {
    maxOutputTokens: 10_000,
    maxInputTokens: 300_000,
    cacheSupport: ['system'],
    toolChoiceSupport: ['auto'],
  },
  opus4: {
    maxOutputTokens: 32_000,
    maxInputTokens: 200_000,
    cacheSupport: ['system', 'message', 'tool'],
    reasoningSupport: true,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
  sonnet4: {
    maxOutputTokens: 64_000,
    maxInputTokens: 200_000,
    cacheSupport: ['system', 'message', 'tool'],
    reasoningSupport: true,
    toolChoiceSupport: ['any', 'auto', 'tool'],
  },
};

export const bedrockConverse = async (
  workerId: string,
  modelTypes: ModelType[],
  input: Omit<ConverseCommandInput, 'modelId'>,
  maxTokensExceededCount = 0
) => {
  if (maxTokensExceededCount > 5) {
    throw new Error(`Max tokens exceeded too many times (${maxTokensExceededCount})`);
  }
  try {
    const modelOverride = modelTypeSchema
      .optional()
      // empty string to undefined
      .parse(process.env.MODEL_OVERRIDE ? process.env.MODEL_OVERRIDE : undefined);
    const modelType = modelOverride || chooseRandom(modelTypes);
    const { client, modelId, awsRegion, account } = await getModelClient(modelType);
    console.log(`Using ${JSON.stringify({ modelId, awsRegion, account, roleName })}`);
    const command = new ConverseCommand(
      preProcessInput(
        {
          ...input,
          modelId,
        },
        modelType,
        maxTokensExceededCount
      )
    );
    const response = await client.send(command);

    // Track token usage for analytics
    await trackTokenUsage(workerId, modelId, response);

    return response;
  } catch (error) {
    if (error instanceof ThrottlingException) {
      // Rotate to next account
      const previousIndex = currentAccountIndex;
      currentAccountIndex = (currentAccountIndex + 1) % awsAccounts.length;
      console.log(
        `AWS account ${awsAccounts[previousIndex]} has been throttled. Switching to ${awsAccounts[currentAccountIndex]}.`
      );
    }
    throw error; // Re-throw for handling by upper-level pRetry
  }
};

const shouldUltraThink = (input: ConverseCommandInput): boolean => {
  // Get the last user message to look for keywords
  const messages = input.messages || [];
  const lastUserMessage = messages
    .filter((message) => message.role === 'user' && message.content?.some((c) => c.text != null))
    .pop();
  if (!lastUserMessage?.content) {
    return false;
  }

  // Convert all content parts to string if possible to check for keywords
  const messageText = lastUserMessage.content
    .map((content) => ('text' in content ? content.text : ''))
    .join(' ')
    .toLowerCase();

  // Check for the keywords to adjust thinking budget
  return messageText.includes(ULTRA_THINKING_KEYWORD);
};

const preProcessInput = (input: ConverseCommandInput, modelType: ModelType, maxTokensExceededCount: number) => {
  const modelConfig = modelConfigSchema.parse(modelConfigs[modelType]);
  // we cannot use JSON.parse(JSON.stringify(input)) here because input sometimes contains Buffer object for image.
  input = structuredClone(input);

  // remove toolChoice if not supported
  if (input.toolConfig?.toolChoice) {
    if (modelConfig.toolChoiceSupport.every((choice) => !(choice in input.toolConfig!.toolChoice!))) {
      input.toolConfig.toolChoice = undefined;
    }
  }

  // set maximum number of output tokens
  const adjustedMaxToken = Math.min(modelConfig.maxOutputTokens, defaultOutputTokenCount * 2 ** maxTokensExceededCount);
  input.inferenceConfig = { ...input.inferenceConfig, maxTokens: adjustedMaxToken };

  // enable or disable reasoning
  let enableReasoning = false;
  if (modelConfig.reasoningSupport) {
    if (input.toolConfig?.toolChoice != null) {
      // toolChoice and reasoning cannot be enabled at the same time
    } else if (
      input.messages?.at(-2)?.content?.at(0)?.reasoningContent == null &&
      input.messages?.at(-2)?.content?.at(-1)?.toolUse != null
    ) {
      // reasoning cannot be enabled when the last message is toolUse and toolUse does not have reasoning block.
    } else {
      enableReasoning = true;
    }
  }

  if (enableReasoning) {
    // Detect if we need to adjust the thinking budget based on keywords
    const enableUltraThink = shouldUltraThink(input);
    const budget = enableUltraThink ? Math.min(Math.floor(modelConfig.maxOutputTokens / 2), 31999) : 2000;

    // Apply thinking budget settings
    input.additionalModelRequestFields = {
      reasoning_config: {
        type: 'enabled',
        budget_tokens: budget,
      },
    };

    // Adjust output tokens as well
    input.inferenceConfig = {
      ...input.inferenceConfig,
      maxTokens: Math.max(adjustedMaxToken, Math.min(budget * 2, modelConfig.maxOutputTokens)),
    };
  } else {
    // when we disable reasoning, we have to remove
    // reasoningContent blocks from all the previous message contents
    input.messages = input.messages?.map((message) => {
      message.content = message.content?.filter((c) => {
        return !('reasoningContent' in c);
      });
      return message;
    });
  }
  // remove cachePoints if not supported
  if (!modelConfig.cacheSupport.includes('system') && input.system) {
    for (let i = input.system.length - 1; i >= 0; i--) {
      if ('cachePoint' in input.system[i]) {
        input.system.splice(i, 1);
      }
    }
  }
  if (!modelConfig.cacheSupport.includes('tool') && input.toolConfig?.tools) {
    for (let i = input.toolConfig.tools.length - 1; i >= 0; i--) {
      if ('cachePoint' in input.toolConfig.tools[i]) {
        input.toolConfig.tools.splice(i, 1);
      }
    }
  }
  if (!modelConfig.cacheSupport.includes('message') && input.messages) {
    for (const message of input.messages) {
      const content = message.content;
      if (!content) continue;
      for (let i = content.length - 1; i >= 0; i--)
        if ('cachePoint' in content[i]) {
          content.splice(i, 1);
        }
    }
  }

  return input;
};

const getModelClient = async (modelType: ModelType) => {
  const { awsRegion, modelId } = chooseModelAndRegion(modelType);
  const account = awsAccounts[currentAccountIndex];

  if (awsAccounts.length === 0 || !account) {
    return { client: new BedrockRuntimeClient({ region: awsRegion }), modelId };
  }

  const cred = await getCredentials(account);
  const client = new BedrockRuntimeClient({
    region: awsRegion,
    credentials: {
      accessKeyId: cred.AccessKeyId!,
      secretAccessKey: cred.SecretAccessKey!,
      sessionToken: cred.SessionToken!,
    },
  });
  return { client, modelId, awsRegion, account };
};

const chooseRandom = <T>(choices: T[]) => {
  return choices[Math.floor(Math.random() * choices.length)];
};

const chooseModelAndRegion = (modelType: ModelType) => {
  const availableRegions = ['us'];
  const region = chooseRandom(availableRegions);
  let awsRegion = 'us-west-2';
  if (region == 'eu') awsRegion = 'eu-west-1';
  if (region == 'apac') awsRegion = 'ap-northeast-1';
  let modelId = '';
  switch (modelType) {
    case 'sonnet3.5v1':
      modelId = 'anthropic.claude-3-5-sonnet-20240620-v1:0';
      break;
    case 'sonnet3.5':
      modelId = 'anthropic.claude-3-5-sonnet-20241022-v2:0';
      break;
    case 'sonnet3.7':
      modelId = 'anthropic.claude-3-7-sonnet-20250219-v1:0';
      break;
    case 'haiku3.5':
      modelId = 'anthropic.claude-3-5-haiku-20241022-v1:0';
      break;
    case 'nova-pro':
      modelId = 'amazon.nova-pro-v1:0';
      break;
    case 'opus4':
      modelId = 'anthropic.claude-opus-4-20250514-v1:0';
      break;
    case 'sonnet4':
      modelId = 'anthropic.claude-sonnet-4-20250514-v1:0';
      break;
    default:
      throw new Error(`Unknown model type: ${modelType}`);
  }
  modelId = `${region}.${modelId}`;
  return {
    modelId,
    awsRegion,
  };
};

const getCredentials = async (account: string) => {
  const roleArn = `arn:aws:iam::${account}:role/${roleName}`;
  const res = await sts.send(
    new AssumeRoleCommand({
      RoleArn: roleArn,
      RoleSessionName: 'remote-swe-session',
    })
  );
  if (!res.Credentials) {
    throw new Error('No credentials returned from assumeRole');
  }
  return res.Credentials;
};

const trackTokenUsage = async (workerId: string, modelId: string, response: ConverseResponse) => {
  if (!TableName) {
    return;
  }
  if (!response.usage) {
    console.warn('No usage information in response');
    return;
  }

  const { inputTokens, outputTokens, cacheReadInputTokens, cacheWriteInputTokens } = response.usage;

  // Retrieve or create item with PK: token-<workerId>, SK: modelId
  try {
    // Get existing item if available (ignoring race condition for brevity)
    const existingItem = await ddb.send(
      new GetCommand({
        TableName,
        Key: {
          PK: `token-${workerId}`,
          SK: modelId,
        },
      })
    );

    if (existingItem.Item) {
      // Update (increment token counts) if item exists
      await ddb.send(
        new UpdateCommand({
          TableName,
          Key: {
            PK: `token-${workerId}`,
            SK: modelId,
          },
          UpdateExpression:
            'ADD inputToken :inputTokens, outputToken :outputTokens, cacheReadInputTokens :cacheReadInputTokens, cacheWriteInputTokens :cacheWriteInputTokens',
          ExpressionAttributeValues: {
            ':inputTokens': inputTokens || 0,
            ':outputTokens': outputTokens || 0,
            ':cacheReadInputTokens': cacheReadInputTokens || 0,
            ':cacheWriteInputTokens': cacheWriteInputTokens || 0,
          },
        })
      );
    } else {
      // Create new item if it doesn't exist yet
      await ddb.send(
        new PutCommand({
          TableName,
          Item: {
            PK: `token-${workerId}`,
            SK: modelId,
            inputToken: inputTokens || 0,
            outputToken: outputTokens || 0,
            cacheReadInputTokens: cacheReadInputTokens || 0,
            cacheWriteInputTokens: cacheWriteInputTokens || 0,
          },
        })
      );
    }
  } catch (error) {
    // do not throw error to avoid affecting the primary process
    console.error('Error tracking token usage:', error);
  }
};
