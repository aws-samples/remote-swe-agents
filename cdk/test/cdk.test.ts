import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { readFileSync } from 'fs';
import { MainStack } from '../lib/cdk-stack';

test('Snapshot test', () => {
  jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

  const app = new cdk.App({
    context: {
      ...JSON.parse(readFileSync('cdk.json').toString()).context,
    },
  });

  const main = new MainStack(app, `TestMainStack`, {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
    slack: {
      botTokenParameterName: '/remote-swe/slack/bot-token',
      signingSecretParameterName: '/remote-swe/slack/signing-secret',
      adminUserIdList: undefined,
    },
    github: {
      privateKeyParameterName: '/remote-swe/github/app-private-key',
      appId: '123456',
      installationId: '9876543',
    },
    workerAmiIdParameterName: '/remote-swe/worker/ami-id',
  });

  expect(Template.fromStack(main)).toMatchSnapshot();
});
