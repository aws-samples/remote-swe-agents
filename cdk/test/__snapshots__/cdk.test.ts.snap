// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Description": "Remote SWE Agents stack (uksb-lv52f92xel)",
  "Outputs": Object {
    "SlackBoltEndpointUrl89D2B06F": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "SlackBoltApiFA4A2047",
          "ApiEndpoint",
        ],
      },
    },
    "StorageBucketName37AA483C": Object {
      "Value": Object {
        "Ref": "StorageImageBucket99BA9550",
      },
    },
    "StorageTableNameB6E3E5D6": Object {
      "Value": Object {
        "Ref": "StorageHistory251A3AE8",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsservicecanonicalubuntuserver2404stablecurrentamd64hvmebsgp3amiidC96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/canonical/ubuntu/server/24.04/stable/current/amd64/hvm/ebs-gp3/ami-id",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "EC2GarbageCollectorHandler4E42AD35": Object {
      "DependsOn": Array [
        "EC2GarbageCollectorHandlerServiceRole3E1396FE",
      ],
      "Properties": Object {
        "Architecture": "arm64",
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "REDACTED",
        },
        "Environment": Object {
          "Variables": Object {
            "EXPIRATION_IN_DAYS": "1",
            "IMAGE_RECIPE_NAME": Object {
              "Fn::GetAtt": [
                "WorkerImageBuilderImagePipelineV2ImageRecipe351515D4",
                "Name",
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "EC2GarbageCollectorHandlerServiceRole3E1396FE",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EC2GarbageCollectorHandlerServiceRole3E1396FE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EC2GarbageCollectorHandlerServiceRoleDefaultPolicy1FE51FB9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:DeleteSnapshot",
                "ec2:DeregisterImage",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "ec2:TerminateInstances",
                "imagebuilder:DeleteImage",
                "imagebuilder:GetImage",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EC2GarbageCollectorHandlerServiceRoleDefaultPolicy1FE51FB9",
        "Roles": Array [
          Object {
            "Ref": "EC2GarbageCollectorHandlerServiceRole3E1396FE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EC2GarbageCollectorScheduleF6CD97DD": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(2 hours)",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "EC2GarbageCollectorHandler4E42AD35",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "EC2GarbageCollectorScheduleAllowEventRuleTestMainStackEC2GarbageCollectorHandler347502ED890054F4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "EC2GarbageCollectorHandler4E42AD35",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "EC2GarbageCollectorScheduleF6CD97DD",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`
